In [2]:
import pandas as pd

initiate_search = pd.read_csv('initiate_search.csv')
search_result = pd.read_csv('search_result.csv')
click_search = pd.read_csv('click_search.csv')
search_query = pd.read_csv('search_query.csv')
user_metadata = pd.read_csv('user_metadata.csv')

# First look at the data

**initiate_search**: an event that is triggered when a user enters an empty search
query field to begin entering a query. This is the first event of what we define as a
search session.

In [96]:
display(initiate_search.head())
print(initiate_search.info())
print((initiate_search.searchUuid.value_counts().head(10)))

dt client_platform       event_name  user_id       event_ts  \
0  2021-02-06             iOS  initiate_search      225  1612648404589   
1  2021-02-05             iOS  initiate_search       95  1612565307151   
2  2021-02-05             iOS  initiate_search       95  1612564857792   
3  2021-02-05             iOS  initiate_search       95  1612565891804   
4  2021-02-05             iOS  initiate_search       95  1612564856980   

    server_ts        method                        searchUuid  
0  1612648405    clickInBar  1612648404588.040518700386519682  
1  1612565307  deleteSearch   1612565307150.03093909192593003  
2  1612564858  deleteSearch   1612564857791.03093909192593003  
3  1612565892  deleteSearch   1612565891803.03093909192593003  
4  1612564857  searchButton   1612564856977.03093909192593003

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14385 entries, 0 to 14384
Data columns (total 8 columns):
dt                 14385 non-null object
client_platform    14385 non-null object
event_name         14385 non-null object
user_id            14385 non-null int64
event_ts           14385 non-null int64
server_ts          14385 non-null int64
method             14385 non-null object
searchUuid         14385 non-null object
dtypes: int64(3), object(5)
memory usage: 899.2+ KB
None
1612556056230.0171297301277280809       3
1612556043999.0171297301277280809       3
aa15730c-142e-48f8-b6a6-006f87656df8    2
4430c281-aa2b-4a04-a12d-1953c767e6ff    2
94b98718-5cb7-4e9f-8722-61cf805a9aa7    2
78a1e913-f27a-4443-92de-7368d62361bf    2
dea48dc1-0338-4b85-9694-e2e9ceb4ba6f    2
92eca60f-8e2b-47ef-b07a-e6858208cf19    2
31fb49ed-0ad6-428e-9c98-897b4f3bb97b    1
d1d1bb0a-e2c5-4f60-ae45-a34cc398dd13    1
Name: searchUuid, dtype: int64


In [9]:
for column in initiate_search.columns:
    print("---- %s ---" % column)
    print(initiate_search[column].value_counts(), '\n')

---- dt ---
2021-02-06    2551
2021-02-05    2346
2021-02-07    2081
2021-02-01    1912
2021-02-04    1877
2021-02-03    1857
2021-02-02    1761
Name: dt, dtype: int64 

---- client_platform ---
iOS        7238
android    5642
desktop    1505
Name: client_platform, dtype: int64 

---- event_name ---
initiate_search    14385
Name: event_name, dtype: int64 

---- user_id ---
176     225
1076    158
75      150
852     116
866     109
       ... 
835       1
1117      1
819       1
811       1
338       1
Name: user_id, Length: 1198, dtype: int64 

---- event_ts ---
1612145670539    3
1612556090301    3
1612145670540    3
1612556090306    3
1612527836733    2
                ..
1612418395471    1
1612635962709    1
1612377987662    1
1612399140177    1
1612517988539    1
Name: event_ts, Length: 14298, dtype: int64 

---- server_ts ---
1612490936    64
1612734578    64
1612455756    46
1612540456    42
1612523112    41
              ..
1612232866     1
1612447901     1
1612646556     1
161

**search_query**: search_query: the *backend response event* that is used to populate the search
engine results page. Contains lots of metadata pertaining to the SERP.

In [16]:
display(search_query.head())
print(search_query.info())

dt  user_id   server_ts            query  \
0  2021-02-07     1094  1612718227      georg benda   
1  2021-02-07      863  1612726941              in    
2  2021-02-07      850  1612715317              jid   
3  2021-02-07       49  1612725870      Benz window   
4  2021-02-07      231  1612726235  London homesick   

                              queryUuid  \
0  8698c2dc-fd4e-4f2b-81e1-c79601cd5f7a   
1  98ffcc2e-f84e-4587-80fd-89e1b178225d   
2  83a827ae-a93e-4148-88a1-c9f2715b886a   
3  2487a99a-efb9-41ee-8f8d-e4203ea22f4c   
4  a1ad34dd-0927-42cf-84f4-e9c6fafbfb86   

                                        trackResults  \
0  [{"id":"11363806","popularityWW":"0"},{"id":"1...   
1  [{"id":"134858528","popularityCountry":"97"},{...   
2  [{"id":"108769424","popularityCountry":"64"},{...   
3  [{"id":"102847894","popularityWW":"0"},{"id":"...   
4  [{"id":"35717236","popularityCountry":"4"},{"i...   

                                       artistResults  \
0  [{"id":"12388224","popularityWW":"0"},{"id":"2...   
1  [{"id":"24053","popularityCountry":"48"},{"id"...   
2  [{"id":"7061089","popularityCountry":"92"},{"i...   
3             [{"id":"11408668","popularityWW":"0"}]   
4                                                 []   

                                        albumResults  \
0  [{"id":"165946610","popularityCountry":"2"},{"...   
1  [{"id":"52885530","popularityCountry":"90"},{"...   
2  [{"id":"100122410","popularityCountry":"76"},{...   
3                                                 []   
4                                                 []   

                                        videoResults  \
0                                                 []   
1  [{"id":"37690657","popularityCountry":"62"},{"...   
2  [{"id":"120672680","popularityCountry":"39"},{...   
3                                                 []   
4                                                 []   

                                        responseTime  \
0  {"track":16,"video":4,"artist":18,"album":10,"...   
1  {"track":92,"video":16,"artist":74,"album":39,...   
2  {"track":9,"video":6,"artist":11,"album":6,"pl...   
3  {"track":35,"video":4,"artist":34,"album":10,"...   
4  {"track":71,"video":14,"artist":32,"album":33,...   

                                          topResults  
0  [{"id":165946610,"contentType":"Album"},{"id":...  
1  [{"id":134858528,"contentType":"Track"},{"id":...  
2  [{"id":7061089,"contentType":"Artist"},{"id":1...  
3  [{"id":46783326,"contentType":"Playlist"},{"id...  
4  [{"id":7896962,"contentType":"Playlist"},{"id"...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27800 entries, 0 to 27799
Data columns (total 11 columns):
dt               27800 non-null object
user_id          27800 non-null int64
server_ts        27800 non-null int64
query            27800 non-null object
queryUuid        27800 non-null object
trackResults     27800 non-null object
artistResults    27800 non-null object
albumResults     27800 non-null object
videoResults     27800 non-null object
responseTime     27800 non-null object
topResults       27800 non-null object
dtypes: int64(2), object(9)
memory usage: 2.3+ MB
None


**search_result**: an event that is triggered when query results are received after a
user performs a query

In [98]:
display(search_result.head())
print(search_result.info())
print(search_result.queryUuid.value_counts())

dt client_platform     event_name  user_id       event_ts  \
0  2021-02-06             iOS  search_result      250  1612639179188   
1  2021-02-06             iOS  search_result      987  1612643266165   
2  2021-02-06             iOS  search_result      266  1612649431495   
3  2021-02-06         android  search_result      773  1612644037697   
4  2021-02-06         android  search_result      773  1612644463249   

    server_ts            query                            searchUuid  \
0  1612639179              The      1612639176586.041507019172635133   
1  1612643266                S     1612643264805.0175305242380418880   
2  1612649431    Full of smoke      1612649421092.042425692113708085   
3  1612644069  microphones in   0b61e3c2-091e-4273-b15a-e3e80326a81d   
4  1612644488         blue fla  72433904-257a-4899-ad11-cafca503381a   

                              queryUuid  
0  2574e1b5-023f-49e7-bfca-94bc1630fa57  
1  4c9dea6e-069f-4f56-8f98-9a354d498f46  
2  c27b2912-b7ba-4642-b663-83ebba4bca7d  
3  015c5ac7-1d33-4788-95eb-bed1cd1dd422  
4  a8032920-6fe6-4e34-8b7d-860ab9cbeb75

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31449 entries, 0 to 31448
Data columns (total 9 columns):
dt                 31449 non-null object
client_platform    31449 non-null object
event_name         31449 non-null object
user_id            31449 non-null int64
event_ts           31449 non-null int64
server_ts          31449 non-null int64
query              31424 non-null object
searchUuid         31449 non-null object
queryUuid          31424 non-null object
dtypes: int64(3), object(6)
memory usage: 2.2+ MB
None
offline                                 42
543c49d9-fb71-429c-9ad7-cfc173b285a2     9
165e39a2-94d7-44bd-aa9e-fbe08073d6b3     8
392ce0c7-2dad-4b26-9976-cf9c16f019b4     8
ada61a29-cd51-4c87-936c-bdd58e529fb4     7
                                        ..
923d0b1f-3435-4823-a61f-8b98687e278a     1
b9d0e88d-6802-4df9-bdf0-9ddb7a1979f4     1
f81af540-7b49-49a0-9c01-1b68a3ea4095     1
cc209b76-c8c4-436c-ba7f-b1284b1abdb4     1
4a6749a7-91ff-4632-bc4c-feaf4d0d3605     

A few null values in query and queryUUid: results with no query? Can join with initiate_search on searchUuid

Can join with search_query with search_result on queryUuid.

**click_search**: an event that is triggered when a user makes a click on the search
page (e.g. clicking on a search result). We typically consider a click on content
the *successful completion* of a search session.

In [17]:
display(click_search.head())
print(click_search.info())

dt client_platform    event_name  user_id       event_ts  \
0  2021-02-02         android  click_search      951  1612233213601   
1  2021-02-02         android  click_search      951  1612233225774   
2  2021-02-02             iOS  click_search     1128  1612233621630   
3  2021-02-02             iOS  click_search      177  1612233483273   
4  2021-02-02             iOS  click_search      177  1612232890509   

    server_ts                            searchUuid  \
0  1612233291  bb45102e-0efe-45e3-8298-c74728c0dcb4   
1  1612233291  bb45102e-0efe-45e3-8298-c74728c0dcb4   
2  1612233621     1612233604129.0176685845395195847   
3  1612233483      1612232865985.036730574371815387   
4  1612232891      1612232865985.036730574371815387   

                              queryUuid  pageId     endResult contentType  \
0                                   NaN  search  cancelSearch         NaN   
1  5fa761dd-f937-4431-903d-cc2d31b8b5bd  search      playback       track   
2  f6355ee9-e522-4a12-8839-3bba00e561fb  search      playback       track   
3  49a5b5d3-1cec-48cd-8686-b1abe942e66e  search    navigation       album   
4  49a5b5d3-1cec-48cd-8686-b1abe942e66e  search    navigation      artist   

   contentId  contentPlacement          moduleId buttonId  
0        NaN               NaN               NaN  xSearch  
1  109582305               1.0  search_topresult      NaN  
2   70170184               1.0  search_topresult      NaN  
3  160351543               4.0  search_topresult      NaN  
4    4332277               0.0  search_topresult      NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16717 entries, 0 to 16716
Data columns (total 15 columns):
dt                  16717 non-null object
client_platform     16717 non-null object
event_name          16717 non-null object
user_id             16717 non-null int64
event_ts            16717 non-null int64
server_ts           16717 non-null int64
searchUuid          16717 non-null object
queryUuid           15681 non-null object
pageId              13476 non-null object
endResult           16717 non-null object
contentType         14845 non-null object
contentId           14441 non-null object
contentPlacement    14634 non-null float64
moduleId            14914 non-null object
buttonId            1753 non-null object
dtypes: float64(1), int64(3), object(11)
memory usage: 1.9+ MB
None


In [19]:
display(user_metadata.head())
print(user_metadata.info())

user_id    birthday gender countryCode   free  churned subscriptionType
0      730  1993-04-29    NaN          US  False    False     PREMIUM_PLUS
1      338         NaN    NaN          DE  False    False          PREMIUM
2      435         NaN    NaN          DE  False    False     PREMIUM_PLUS
3      829  1998-12-30      f          US  False    False          PREMIUM
4     1075         NaN      m          DE  False    False     PREMIUM_PLUS

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 7 columns):
user_id             1200 non-null int64
birthday            601 non-null object
gender              387 non-null object
countryCode         1200 non-null object
free                1200 non-null bool
churned             1200 non-null bool
subscriptionType    1200 non-null object
dtypes: bool(2), int64(1), object(4)
memory usage: 49.3+ KB
None


____
# metrics 
Let's now work throught the suggested metrics to start off:

## Number of search sessions with a result

For this I'll do an outer join the initiate_search df with search results df on the "searchUuid". An outer join will give us initiated searches linked to any front end search results as well as searches with no result which I am curious about. As there are numerous rows for unique searchUuids, I will group by those.

In [100]:
search_sesh = pd.merge(initiate_search, search_result, on='searchUuid', how='outer')
search_sesh

dt_x client_platform_x     event_name_x  user_id_x    event_ts_x  \
0      2021-02-06               iOS  initiate_search      225.0  1.612648e+12   
1      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
2      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
3      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
4      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
...           ...               ...              ...        ...           ...   
37778         NaN               NaN              NaN        NaN           NaN   
37779         NaN               NaN              NaN        NaN           NaN   
37780         NaN               NaN              NaN        NaN           NaN   
37781         NaN               NaN              NaN        NaN           NaN   
37782         NaN               NaN              NaN        NaN           NaN   

        server_ts_x        method                            searchUuid  \
0      1.612648e+09    clickInBar      1612648404588.040518700386519682   
1      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
2      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
3      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
4      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
...             ...           ...                                   ...   
37778           NaN           NaN  97932ffb-5193-4389-9283-f59e716c8319   
37779           NaN           NaN  0fe5356f-f021-4e63-9bea-436b146d768c   
37780           NaN           NaN  ef7ecf94-e201-4d15-925e-324945b62eb3   
37781           NaN           NaN  30f5c352-df15-49c9-b798-b7569072b257   
37782           NaN           NaN  5b9dca6d-c25b-4eef-88cf-2480f311de91   

             dt_y client_platform_y   event_name_y  user_id_y    event_ts_y  \
0             NaN               NaN            NaN        NaN           NaN   
1      2021-02-05               iOS  search_result       95.0  1.612565e+12   
2      2021-02-05               iOS  search_result       95.0  1.612565e+12   
3      2021-02-05               iOS  search_result       95.0  1.612565e+12   
4      2021-02-05               iOS  search_result       95.0  1.612565e+12   
...           ...               ...            ...        ...           ...   
37778  2021-02-07           desktop  search_result      533.0  1.612721e+12   
37779  2021-02-07           desktop  search_result      533.0  1.612721e+12   
37780  2021-02-07           desktop  search_result      533.0  1.612721e+12   
37781  2021-02-07           desktop  search_result       29.0  1.612725e+12   
37782  2021-02-07           desktop  search_result       39.0  1.612721e+12   

        server_ts_y                          query  \
0               NaN                            NaN   
1      1.612565e+09                         Moonli   
2      1.612565e+09                            Moo   
3      1.612565e+09                              M   
4      1.612565e+09                      Moonlight   
...             ...                            ...   
37778  1.612721e+09  stereo dancer absolute reason   
37779  1.612721e+09  stereo dancer absolute reason   
37780  1.612721e+09  stereo dancer absolute reason   
37781  1.612725e+09                            ed    
37782  1.612721e+09               cocorosie we are   

                                  queryUuid  
0                                       NaN  
1      7ece7b7a-8443-4860-b6fb-f8b3408ff18f  
2      31c71b64-bee1-4e0f-9d1e-e21b60ecd9a7  
3      ab089736-7466-4e9e-b8b1-71f5868d28dc  
4      6e2df46c-b442-446f-84ea-51432dfc4531  
...                                     ...  
37778  3bb16988-d1fc-4e20-8c0f-fa82a4b42863  
37779  3bb16988-d1fc-4e20-8c0f-fa82a4b42863  
37780  3bb16988-d1fc-4e20-8c0f-fa82a4b42863  
37781  1c028884-86c6-401f-8467-826a64e6336f  
37782  321ec47b-53ac-44b7-8ce2-0c0a

Interesting, there are a higher number of rows expected here. It appears that there are results without the initial searches.

In [27]:
print(search_sesh.event_name_x.value_counts(dropna=False), '\n')
print(search_sesh.method.value_counts(dropna=False), '\n')
print(search_sesh.event_name_y.value_counts(dropna=False))

initiate_search    37047
NaN                  736
Name: event_name_x, dtype: int64 

clickInBar      19543
deleteSearch    13856
searchButton     2825
directType        822
NaN               736
queueSource         1
Name: method, dtype: int64 

search_result    31467
NaN               6316
Name: event_name_y, dtype: int64


Since a initiate_search event is triggered by any instance that someone enters an empty search query field to begin, it doesn't mean they actually searched. But we are looking for any session with a result.

*There are 6316 initiated searches with no results.*

In [81]:
#creating new table of just instances with searches initiated that also have search results
searches_w_results = search_sesh.loc[((search_sesh['event_name_x'] == 'initiate_search') & 
                                      (search_sesh['event_name_y'] == 'search_result'))] 

searches_w_results

dt_x client_platform_x     event_name_x  user_id_x    event_ts_x  \
1      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
2      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
3      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
4      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
5      2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
...           ...               ...              ...        ...           ...   
37041  2021-02-05               iOS  initiate_search      358.0  1.612509e+12   
37042  2021-02-05               iOS  initiate_search      358.0  1.612509e+12   
37044  2021-02-07           desktop  initiate_search      909.0  1.612682e+12   
37045  2021-02-02               iOS  initiate_search      269.0  1.612241e+12   
37046  2021-02-02               iOS  initiate_search      269.0  1.612241e+12   

        server_ts_x        method                            searchUuid  \
1      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
2      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
3      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
4      1.612565e+09  deleteSearch       1612565307150.03093909192593003   
5      1.612565e+09  deleteSearch       1612564857791.03093909192593003   
...             ...           ...                                   ...   
37041  1.612509e+09  deleteSearch       1612508879947.05073037036991279   
37042  1.612509e+09  deleteSearch       1612508879947.05073037036991279   
37044  1.612682e+09    directType  ea7b84e6-8c2d-4c18-a8d3-b99fef9e2886   
37045  1.612241e+09    clickInBar      1612240786073.042467332385034841   
37046  1.612241e+09    clickInBar      1612240786073.042467332385034841   

             dt_y client_platform_y   event_name_y  user_id_y    event_ts_y  \
1      2021-02-05               iOS  search_result       95.0  1.612565e+12   
2      2021-02-05               iOS  search_result       95.0  1.612565e+12   
3      2021-02-05               iOS  search_result       95.0  1.612565e+12   
4      2021-02-05               iOS  search_result       95.0  1.612565e+12   
5      2021-02-05               iOS  search_result       95.0  1.612565e+12   
...           ...               ...            ...        ...           ...   
37041  2021-02-05               iOS  search_result      358.0  1.612509e+12   
37042  2021-02-05               iOS  search_result      358.0  1.612509e+12   
37044  2021-02-07           desktop  search_result      909.0  1.612682e+12   
37045  2021-02-02               iOS  search_result      269.0  1.612241e+12   
37046  2021-02-02               iOS  search_result      269.0  1.612241e+12   

        server_ts_y             query                             queryUuid  
1      1.612565e+09            Moonli  7ece7b7a-8443-4860-b6fb-f8b3408ff18f  
2      1.612565e+09               Moo  31c71b64-bee1-4e0f-9d1e-e21b60ecd9a7  
3      1.612565e+09                 M  ab089736-7466-4e9e-b8b1-71f5868d28dc  
4      1.612565e+09         Moonlight  6e2df46c-b442-446f-84ea-51432dfc4531  
5      1.612565e+09  Feel good aslove  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  
...             ...               ...                                   ...  
37041  1.612509e+09              Arlo  70d7ca3d-a93e-4062-9588-9adfa814a778  
37042  1.612509e+09          Arlo par  0041bc7c-03f2-4fbb-8a88-338047e6fc64  
37044  1.612682e+09                 e  50f84a76-177c-4351-9095-a69d8c7ebf8c  
37045  1.612241e+09              Kumb  10b9cf6b-fa81-4904-b4a8-8af095227126  
37046  1.612241e+09               Kum  15e1eb96-4bfb-439e-a11c-17415604605b  

[30731 rows x 16 columns]

Number of search *events* with a search result: 30,731

If we look closely, we can see that there are numerous *events* for many unique searchUuids.
I will assume a search *session* included the various searches within the session. The amount of unique 'searchUuids' will tell us how many search sessions there are.

In [86]:
len(searches_w_results['searchUuid'].unique())


8064

**There are 8064 (unique) search "sessions" that returned results.**


Let's investigate the different methods that returned results.

In [48]:
searches_w_results.method.value_counts()

clickInBar      16798
deleteSearch    11636
searchButton     1791
directType        506
Name: method, dtype: int64

andn the methods where initiated search actions didn't return results.

In [54]:
search_sesh.loc[(search_sesh['event_name_y'] != 'search_result')].method.value_counts() 

clickInBar      2745
deleteSearch    2220
searchButton    1034
directType       316
queueSource        1
Name: method, dtype: int64

##  Number of search sessions with a click per user

"per user" means we need to group by user. How many search sessions with a click does each user have.

Note: there are 16717 events of click_search

For this I will join the click_search df to the searches_w_results df.

In [130]:
#the following inner join will give us the total number of sessions with a click
search_reswclick = pd.merge(searches_w_results, click_search, on='searchUuid', how='inner') 
pd.set_option('display.max_columns', None)
search_reswclick.head()

dt_x client_platform_x     event_name_x  user_id_x    event_ts_x  \
0  2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
1  2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
2  2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
3  2021-02-05               iOS  initiate_search       95.0  1.612565e+12   
4  2021-02-05               iOS  initiate_search       95.0  1.612566e+12   

    server_ts_x        method                       searchUuid        dt_y  \
0  1.612565e+09  deleteSearch  1612564857791.03093909192593003  2021-02-05   
1  1.612565e+09  deleteSearch  1612564857791.03093909192593003  2021-02-05   
2  1.612565e+09  deleteSearch  1612564857791.03093909192593003  2021-02-05   
3  1.612565e+09  deleteSearch  1612564857791.03093909192593003  2021-02-05   
4  1.612566e+09  deleteSearch  1612565891803.03093909192593003  2021-02-05   

  client_platform_y   event_name_y  user_id_y    event_ts_y   server_ts_y  \
0               iOS  search_result       95.0  1.612565e+12  1.612565e+09   
1               iOS  search_result       95.0  1.612565e+12  1.612565e+09   
2               iOS  search_result       95.0  1.612565e+12  1.612565e+09   
3               iOS  search_result       95.0  1.612565e+12  1.612565e+09   
4               iOS  search_result       95.0  1.612566e+12  1.612566e+09   

              query                           queryUuid_x          dt  \
0  Feel good aslove  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  2021-02-05   
1    Feel good aslo  de6c857f-dfff-4d47-986b-e6604ceb6c28  2021-02-05   
2         Feel good  6449a0ee-d247-446c-b0cb-0d87ffaa680a  2021-02-05   
3                 F  52086da0-4cc4-4e9b-bd77-b6e0615c2e07  2021-02-05   
4               Sus  d4372bf3-9763-4731-8724-4298ebeaa3e9  2021-02-05   

  client_platform    event_name  user_id       event_ts   server_ts  \
0             iOS  click_search       95  1612564874244  1612564874   
1             iOS  click_search       95  1612564874244  1612564874   
2             iOS  click_search       95  1612564874244  1612564874   
3             iOS  click_search       95  1612564874244  1612564874   
4             iOS  click_search       95  1612566280533  1612566280   

                            queryUuid_y  pageId   endResult contentType  \
0  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  search    playback       track   
1  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  search    playback       track   
2  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  search    playback       track   
3  26a451cd-31c8-42e3-a88a-b66a7cf7be5a  search    playback       track   
4  d4372bf3-9763-4731-8724-4298ebeaa3e9  search  navigation      artist   

  contentId  contentPlacement          moduleId buttonId  
0  50880918               0.0  search_topresult      NaN  
1  50880918               0.0  search_topresult      NaN  
2  50880918               0.0  search_topresult      NaN  
3  50880918               0.0  search_topresult      NaN  
4   3520799               0.0  search_topresult      NaN

Note: there are now 47082 rows with our inner join. This is because the click_search df has a row(click) for each click of unique content. However when we join, there is the same click for multiple queries. 

For example: the same search_topresult of "contentPlacement" position 0.0 appears for various querys from the same string "query" (Feel goodaslove, Feel goodalso, Feel good, F). It is the same click for each. This makes me think that the search bar in TIDAL app is active, where searches continuously appear as you are typing the query in.

So, there are 47082 search events that resulted in a click.

Since I've already defined search session as those with unique searchUuids, I'll have to create a df with just unique searchUuids and then group by the user.

In [120]:
len(search_reswclick.searchUuid.unique())

6635

In [126]:
len(click_search.searchUuid.unique())

8753

Note: there appear to be click searches without initiated_searches

**Answer below:**

In [147]:
#we have clicks for each single event already, now we want the number of unique search sessions per user
clicks_per_user = search_reswclick.groupby("user_id")["searchUuid"].count()

#expand to see all rows:
pd.set_option('display.max_rows', None)

clicks_per_user.to_frame()

searchUuid
user_id            
1                 3
2                 6
3                28
4                78
5                15
6                32
7                 5
8                54
9                13
10               24
11              221
12               53
13               35
14               33
15                4
16                7
17               67
18              174
19              117
20               32
21               26
22               38
23               26
24              338
25                1
26               82
27                2
28               69
29               85
30               95
31               72
33               48
34               11
35               24
36                3
37               38
39               18
40               87
41               13
42              160
43                4
44               15
45                3
46                4
47                2
48              147
49               15
50                2
51                7
52               10
53               46
54               45
55              130
56               56
57                8
58               62
59               61
60               35
61               31
62                5
63               82
64                1
65                9
66              110
67               27
68                3
69               19
70                4
71               12
72               45
73               20
74               21
75               34
76                3
77               45
80                2
81               70
82              150
83                8
84               12
85                5
86                3
87                6
88               23
89              175
90               28
91               26
92               44
94               67
95               37
96                7
97              120
98               69
99                1
100              32
101              63
102               4
103               9
104              44
105              17
106              54
107              10
108             100
109              51
110              13
111               1
112              23
114              15
115              11
116              21
117              18
118              18
119               3
120               7
121              16
122              10
123              30
124              39
125              10
126               2
127              75
128              21
129              27
130              41
131              48
132             316
133              45
134               5
135              11
136              10
137              31
138              90
139              12
140              15
141              32
142              17
143               3
145              73
146              12
147              83
148               4
149              69
151               4
152               1
153              20
154              14
155               9
156              24
157              30
158              27
159              14
160               3
161               9
162              12
163              37
164              16
165              49
166              76
167              10
168             100
169              30
170              46
171              28
172              22
173              20
174              15
175              12
176              13
177             197
178              10
179              89
180              30
181               2
182              23
183               4
184              66
185              30
186              19
187              10
188              30
189              74
190               5
191              21
192              54
193              21
194               4
195              61
196              13
197               4
198              26
199              83
200              92
201              81
202              18
203               3
204               3
205              51
206             202
207      

## Distribution of clicks by content type

IOW: How many clicks did each content type get?

In [148]:
click_search

KeyboardInterrupt: 